### PR TITLE
Add support for IBM threaded compilers, xl*_r

### PR DESCRIPTION
**Pull request by serbanmaerean**
Added new compiler class, xl_r; added default flags to the compilers.yaml file.

### CODE REVIEW

**Comment by davydden** [comment]
for a reference, this is a part of https://github.com/LLNL/spack/pull/2861

p.s. @serbanmaerean i think it's good idea to merge compiler first and then focus on adding patches from https://github.com/LLNL/spack/pull/2861 👍

**Comment by serbanmaerean** [comment]
@davydden I want to do this in small pieces that can be easily verified.  I am starting with the compiler changes and, once approved and merged, I'll continue with the changes for Spack packages.  I want to create a few pull requests, maybe not one for each package file, but keep each request small and manageable.

**Comment by eklee15** [comment]
@davydden  FYI, we are planning on modifying patches and packages for xl_r from the latest version of spack. The following packages will be modified first. 

cmake
netlib-lapack
netlib-scalapack
fftw
matio
boost
hdf5
metis
parmetis
netcdf
cuite-sparse
superlu-dist
mumps
hypre
trilinos
petsc
magma

File: lib/spack/spack/compilers/__init__.py [hunk 1]
```diff
@@ -40,6 +40,7 @@
 
 _imported_compilers_module = 'spack.compilers'
 _path_instance_vars = ['cc', 'cxx', 'f77', 'fc']
+_flags_instance_vars = ['cflags', 'cxxflags', 'fflags']
```
Comment (becker33) [inline lib/spack/spack/compilers/__init__.py]: These should probably include `cppflags` as well, for completeness's sake. I understand why we wouldn't want to add `ldflags` or `ldlibs`.

Are the fortran flags you add in the xl_r compiler absolutely necessary for it to run properly? If not, this should probably be changed in favor of some optional method. We don't want to restrict how spack users configure their compilers any more than necessary.

Comment (serbanmaerean) [inline lib/spack/spack/compilers/__init__.py]: The fflags are required by the xlf* compiler to work properly.  I see now the following definitions in spec.py:
_valid_compiler_flags = [
    'cflags', 'cxxflags', 'fflags', 'ldflags', 'ldlibs', 'cppflags']

I will add cppflags.  Do I do a new commit on my repo and a new pull request?!?

Comment (adamjstewart) [inline lib/spack/spack/compilers/__init__.py]: Nah, new commit on the same pull request.

**Comment by serbanmaerean** [comment]
I need to understand how do I need to proceed from now.  Shall I wait for a merge before adding more commits?  Or shall I add all necessary commits before a full merge?

**Comment by adamjstewart** [comment]
You'll have to fix the flake8 warnings that are causing the tests to fail. But other than that, this PR looks good to merge as is. Best wait until this is merged to start fixing individual packages.

**Comment by serbanmaerean** [comment]
I apologize for my ignorance, what are the flake8 warnings?  How can I see them?

**Comment by adamjstewart** [comment]
You'll notice at the bottom of the page, there is a message that says "Some checks were not successful". If you click on "Details" next to the failing test, it will take you to the Travis CI page. Click on the failing test for the flake8 test suite and at the bottom you'll see the warnings it raised:
```
lib/spack/spack/compilers/__init__.py:65: [E501] line too long (83 > 79 characters)
lib/spack/spack/compilers/xl_r.py:30: [E302] expected 2 blank lines, found 1
lib/spack/spack/compilers/xl_r.py:44: [E201] whitespace after '{'
lib/spack/spack/compilers/xl_r.py:44: [E203] whitespace before ':'
lib/spack/spack/compilers/xl_r.py:45: [E203] whitespace before ':'
lib/spack/spack/compilers/xl_r.py:46: [E203] whitespace before ':'
lib/spack/spack/compilers/xl_r.py:47: [E203] whitespace before ':'
lib/spack/spack/compilers/xl_r.py:47: [E202] whitespace before '}'
lib/spack/spack/compilers/xl_r.py:111: [E203] whitespace before ':'
lib/spack/spack/compilers/xl_r.py:116: [E303] too many blank lines (2)
```

**Comment by tgamblin** [comment]
@serbanmaerean: you can also run those tests locally with `spack flake8` (you'll need the `flake8` package installed in your local python, though).

**Comment by adamjstewart** [comment]
For more information on running Flake8 tests locally, see [the documentation](http://spack.readthedocs.io/en/latest/contribution_guide.html#flake8-tests).

**Comment by serbanmaerean** [comment]
After I installed flake8 with spack (it also built python and py-setuptools among other things), I ran spack flake 8 and received the following error:
[serbanspack@c712f5n13 bin]$ ~/spack/bin/spack flake8
=======================================================
flake8: running flake8 code checks on spack.

Modified files:
  var/spack/repos/builtin/packages/adios/package.py
  var/spack/repos/builtin/packages/cmake/package.py
  var/spack/repos/builtin/packages/spectrum-mpi/package.py
=======================================================
Traceback (most recent call last):
  File "/home/serbanspack/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.8.5/py-flake8-3.0.4-eriomixz2qlogfpvd7xovhe6ehsn467j/bin/flake8", line 5, in <module>
    from pkg_resources import load_entry_point
ImportError: No module named pkg_resources

Flake8 found errors.

I see the following files in spack's py-setuptools directory:
$ ls -l /home/serbanspack/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.8.5/py-setuptools*/bin
total 8
-rwxrwxr-x 1 serbanspack serbanspack 429 Jan 23 11:00 easy_install
-rwxrwxr-x 1 serbanspack serbanspack 437 Jan 23 11:00 easy_install-2.7

I cannot run them--I get same error about failing to import load_entry_point.

What am I missing from the installation?

**Comment by tgamblin** [comment]
`pkg_resources` is part of `setuptools`.  Did you `spack activate py-flake8`?

**Comment by serbanmaerean** [comment]
[serbanspack@c712f5n13 bin]$ ~/spack/bin/spack activate py-flake8
==> Activated extension py-flake8@3.0.4%gcc@4.8.5 arch=linux-rhel7-ppc64le -eriomix for python@2.7.13~tk~ucs4%gcc@4.8.5
[serbanspack@c712f5n13 bin]$ ~/spack/bin/spack flake8
=======================================================
flake8: running flake8 code checks on spack.

Modified files:
  var/spack/repos/builtin/packages/adios/package.py
  var/spack/repos/builtin/packages/cmake/package.py
  var/spack/repos/builtin/packages/spectrum-mpi/package.py
=======================================================
Traceback (most recent call last):
  File "/home/serbanspack/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.8.5/py-flake8-3.0.4-eriomixz2qlogfpvd7xovhe6ehsn467j/bin/flake8", line 5, in <module>
    from pkg_resources import load_entry_point
ImportError: No module named pkg_resources

Flake8 found errors.

**Comment by adamjstewart** [comment]
If you read the "Warning" at the end of the Flake8 documentation section I linked you to, it specifically lists that error message and says that setuptools needs to be in your `PYTHONPATH`. You can either load the module or manually add it to your `PYTHONPATH` to get it working. You can also run `spack activate py-setuptools`

P.S. @tgamblin `spack activate` no longer activates dependencies. See #2760.

**Comment by serbanmaerean** [comment]
Adding py-setuptools to PYTHONPATH didn't do the trick (most probably I did smth wrong.)  I did spack activate py-setuptools and that got it going as I can run spack flake8 now.  I'll have to figure out how to run on the last two commits (by default, it looks at what changed from last commit...)

**Comment by adamjstewart** [comment]
>  I'll have to figure out how to run on the last two commits (by default, it looks at what changed from last commit...)

I think that's because you're working off of your own develop branch. If you had instead created a new branch with just your changes, it would correctly run the tests on all commits since you branched off of develop.

**Comment by adamjstewart** [comment]
Anyway, there's no reason that you _have_ to run flake8 locally. You can just read the warnings from Travis and fix them. Once you push the new commit, you'll see whether or not you missed anything.

**Comment by serbanmaerean** [comment]
I feel more comfortable if I run flake8 locally before I push changes, to avoid unnecessary commits (especially considering that Spack is my first experience w/python and I don't know its syntax rules)

**Comment by adamjstewart** [comment]
No worries. We generally squash all commits in a PR down to a single commit before merging anyway, so don't worry about extraneous commits. But I agree, it is nice to be able to run flake8 locally.

@tgamblin Should we add flake8 and setuptools to Spack's externals? I notice we already have pytest and other optional dependencies there.

**Comment by tgamblin** [comment]
@adamjstewart: I'm not opposed to it, as it's probably useful for committers to be able to run everything out of the box.  I'm tempted, though, to hold off on this and have Spack install it if it's missing instead of including it as an external.

**Comment by serbanmaerean** [comment]
From what I've see thus far, I have not had problems with the `-qfixed` flag in all the packages I built.  It appears to make xlf90 and above accept the fixed form in addition to the free form.  But I will check w/the IBM compiler team and confirm that.

**Comment by tgamblin** [comment]
@serbanmaerean: thanks!  Can you also ask them what their recommendations are on `-qzerosize`, and whether that will disturb a particular Fortran build?   Once we get a verdict on that stuff we can get this merged.

**Comment by serbanmaerean** [comment]
The `-qzerosize` flag is effective only for the Fortran 77 compilers and allows the use of zero size objects.  For Fortran 90 and beyond, it is set by default and has not impact.  It use has no negative side effects.
The `-qstrict` flag allows the Fortran 90+ compilers to parse the source files using fixed form rule.  As a result, if `-qfixed` is in effect, free form files (that are not also fixed form files) will fail to compile regardless of the compiler invocation command.  Users can use the `-qfree` flag in their packages to undo the `-qfixed` flag, as the last one wins.
I tried all different combinations of flags w/the XL Fortran compiler and this is the only way I was able to build Fotran code in the math libraries with Spack.  Besides the code in the math libraries, there are test program from cmake (which we build with gcc) that are compiled with the XL Fortran compiler (when used to build the math libraries) and that throws a wrench in the wheel.

**Comment by tgamblin** [comment]
@serbanmaerean: Ok sounds like we should be using these flags by default!  Thanks for the clarification.  One last thing: can you add this information to the `fflags` method in the compiler class so we don't forget?

**Comment by tgamblin** [comment]
@serbanmaerean: Thanks!

**Comment by serbanmaerean** [comment]
@tgamblin I will start pushing the changes for the package files.  I will start with cmake and move my way up.

**Pull request merged**